### DISCUSSION TITLE
Arduino Portenta H7 Support?

### DISCUSSION THREAD

**Discussion by ajrenzo** [Bodmer/TFT_eSPI#2497] [New processor/display support or capability request] [answered]
I've been trying to find any display driver for the Portenta H7 (especially with touch support) but so far no luck. It doesnt seem like the STM32 processor file already present in TFT_eSPI works despite the H7 having an STM32 processor. Would it be possible to add support for this particular processor? If support already exists, how would I go about setting this up? So far all my efforts have been in vain.

**Comment by Bodmer** [comment]
It should work as a generic processor for SPI displays.

Do NOT use the following line in you setup file (comment it out or delete it) then the generic driver will be used:
#define STM32

**Reply by ajrenzo** [reply]
Thanks for the speedy reply! I commented out the #define STM32 and, while it seems to solve the initial issue I had with that commented in, its showing an error stating that SPI_X was not defined, along with a few others (I only added the final few errors as the log is quite long and seems to be repeating for each processor).

`c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:35: error: 'SPI_X' was not declared in this scope
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                   ^~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:35: note: suggested alternative: 'SPI_6'
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                   ^~~~~
                                   SPI_6

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:24: error: 'spi_get_hw' was not declared in this scope
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                        ^~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:24: note: suggested alternative: 'spi_begin'
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                        ^~~~~~~~~~
                        spi_begin

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:59: error: 'SPI_SSPCR0_DSS_LSB' was not declared in this scope
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                                           ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:59: note: suggested alternative: 'SAI_xCR1_DS_0'
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                                           ^~~~~~~~~~~~~~~~~~
                                                           SAI_xCR1_DS_0

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:79: error: 'SPI_SSPCR0_DSS_BITS' was not declared in this scope
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                                                               ^~~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:79: note: suggested alternative: 'SPI_CR1_SSI'
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
                                                                               ^~~~~~~~~~~~~~~~~~~
                                                                               SPI_CR1_SSI

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3644:7: error: 'hw_write_masked' was not declared in this scope
       hw_write_masked(&spi_get_hw(SPI_X)->cr0, (8 - 1) << SPI_SSPCR0_DSS_LSB, SPI_SSPCR0_DSS_BITS);
       ^~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3650:38: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS){};
                                      ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3650:38: note: suggested alternative: 'SPI_NSS_SOFT'
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS){};
                                      ^~~~~~~~~~~~~~~~~~
                                      SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3657:38: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3657:38: note: suggested alternative: 'SPI_NSS_SOFT'
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~
                                      SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3663:38: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3663:38: note: suggested alternative: 'SPI_NSS_SOFT'
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~
                                      SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3670:38: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3670:38: note: suggested alternative: 'SPI_NSS_SOFT'
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~
                                      SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3684:38: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3684:38: note: suggested alternative: 'SPI_NSS_SOFT'
       while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                      ^~~~~~~~~~~~~~~~~~
                                      SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3693:36: error: 'SPI_SSPSR_BSY_BITS' was not declared in this scope
     while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                    ^~~~~~~~~~~~~~~~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:3693:36: note: suggested alternative: 'SPI_NSS_SOFT'
     while (spi_get_hw(SPI_X)->sr & SPI_SSPSR_BSY_BITS) {};
                                    ^~~~~~~~~~~~~~~~~~
                                    SPI_NSS_SOFT

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp: In member function 'int16_t TFT_eSPI::drawNumber(long int, int32_t, int32_t)':

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5768:3: error: 'ltoa' was not declared in this scope
   ltoa(long_num, str, 10);
   ^~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5768:3: note: suggested alternative: 'utoa'
   ltoa(long_num, str, 10);
   ^~~~
   utoa

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp: In member function 'int16_t TFT_eSPI::drawNumber(long int, int32_t, int32_t, uint8_t)':

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5776:3: error: 'ltoa' was not declared in this scope
   ltoa(long_num, str, 10);
   ^~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5776:3: note: suggested alternative: 'utoa'
   ltoa(long_num, str, 10);
   ^~~~
   utoa

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp: In member function 'int16_t TFT_eSPI::drawFloat(float, uint8_t, int32_t, int32_t, uint8_t)':

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5832:3: error: 'ltoa' was not declared in this scope
   ltoa(temp, str + ptr, 10);
   ^~~~

c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\TFT_eSPI.cpp:5832:3: note: suggested alternative: 'utoa'
   ltoa(temp, str + ptr, 10);
   ^~~~
   utoa`

I'm running Arduino 2.0, and installed using the install .zip function using the zip file from github. I also added the library through the Arduino Library manager prior to doing this in the hopes it would fix the problem. 

Lastly, I noticed the RP2040 was called first during the operation (c:\Users\USER\Documents\Arduino\libraries\TFT_eSPI\Processors/TFT_eSPI_RP2040.c:549:29: error called first), which may be causing the issue. Is there a line I have to add/uncomment to force load the generic driver? I dont see anything in the User_Setup.h file

**Reply by ajrenzo** [reply, marked as answer]
I want to add a quick update to this. I found a previous post referring to SPI_X not being defined: 
https://github.com/Bodmer/TFT_eSPI/issues/1501. Their code compiled when they rolled back their version to 2.3.7. I did the same and the SPI_X didnt show up anymore, so I'm assuming the latest version may have accidentally reintroduced the same issues in 2.4.0. There were some new issues that popped up when I rolled back to 2.3.7 and tried to compile:

'digitalPinToBitMask' was not declared in this scope
'ltoa' was not declared in this scope

It did suggest alternatives, namely changing digitalPinToBitMask to digitalPinToPinName and ltoa to utoa. I went into the TFT_eSPI.cpp and applied those changes, and it does seem to compile now (looks like the Portenta version of Arduino.h uses different macros). I have yet to test it on the Portenta itself but figured this may help with future updates.

UPDATE: It works with the above changes
